edx_exams/apps/api/v1/views.py

```diff
@@ -271,3 +274,116 @@ class ProctoringProvidersView(ListAPIView):
+        except ObjectDoesNotExist:
+            exam_attempt = None
+        return exam_attempt
+
+    @classmethod
+    def set_expiration_window(cls, exam_attempt, default_exp_seconds):
+        """
+        Set exam access ticket expiration window.
+
+        Use default exp window or the time to the end of the exam attempt,
+        whichever is shorter.
+        """
+        secs_remaining = get_exam_attempt_time_remaining(exam_attempt.start_time,
+                                                         exam_attempt.allowed_time_limit_mins)
+
+        # If not past due and attempt time remaining is less than default, set window to remaining
+        if 0 <= secs_remaining < default_exp_seconds:
+            exp_window = secs_remaining
+        else:
+            exp_window = default_exp_seconds
+
+        return exp_window
+
+    @classmethod
+    def get_response(cls, exam, user):
+        """
+        Get response, with exam access ticket if access granted.
+
+        403 error if access is not granted.
+        """
+        claims = {"course_id": exam.course_id, "content_id": exam.content_id}
+        expiration_window = 60
+        exam_attempt = cls.get_current_exam_attempt(user, exam)
+        cookie_name = "exam_access_ticket"
+
+        data = {}
+        grant_access = True
+        response_status = status.HTTP_200_OK
+
+        # If exam attempt exists for user, then grant exam access
+        if exam_attempt is not None:
+            log.info("Creating exam access ticket for attempt")
+            # If started, determine how to set expiration window
+            # todo: should exam attempt status be started or something else?
+            if exam_attempt.status == "started":
+                expiration_window = cls.set_expiration_window(exam_attempt, expiration_window)
+
+        # If exam is past the due date, then grant exam access
+        elif exam.due_date is not None and timezone.now() >= exam.due_date:
+            log.info("Created exam access ticket for exam past due date")
+            response_status = status.HTTP_200_OK
+
+        # Else, do not grant exam access ticket
+        else:
+            log.info("Exam access ticket not granted")
+            grant_access = False
+            data = {"detail": "Exam access ticket not granted"}
+            response_status = status.HTTP_403_FORBIDDEN
+
+        response = Response(status=response_status,
+                            data=data)
+
+        if grant_access:
+            access_ticket = create_jwt(user.lms_user_id, expiration_window, claims)
+            response.set_cookie(cookie_name, access_ticket)
+
+        return response
+
+    def get(self, request, exam_id):
+        """
+        Get exam access ticket as JWT added as cookie to response.
+
+        Exam access ticket corresponds to given exam and user in request.
+        """
+        # todo: Is try/except block necessary? Or does exam always exist?
+        try:
```

Comment:
Question: is a try/except needed here? Is there a case where the exam wouldn't exist in this case?

Reply:
this makes sense to me. I'd suggest a 404 here if the exam id is not found however.

Reply:
Good point.